httpx/concurrency/base.py

```diff
@@ -44,20 +45,98 @@ class BaseTCPStream:
     backends, or for stand-alone test cases.
     """
 
+    def __init__(self, timeout: TimeoutConfig):
+        self.timeout = timeout
+        self.write_buffer = b""
+
     def get_http_version(self) -> str:
         raise NotImplementedError()  # pragma: no cover
 
+    async def read_or_timeout(self, n: int, timeout: typing.Optional[float]) -> bytes:
+        """
+        Read data from the network, or fail in case or timeout.
+
+        Implementations should raise a `WriteTimeout` exception in case of timeout.
+        """
+        raise NotImplementedError()  # pragma: no cover
+
     async def read(
         self, n: int, timeout: TimeoutConfig = None, flag: typing.Any = None
     ) -> bytes:
-        raise NotImplementedError()  # pragma: no cover
+        """
+        Read data from the network.
+
+        Handles common timeout logic, such as switching from read to write timeout
+        and determining the appropriate numeric read timeout value.
+        """
+        if timeout is None:
+            timeout = self.timeout
+
+        while True:
+            # Check our flag at the first possible moment, and use a fine
+            # grained retry loop if we're not yet in read-timeout mode.
+            should_raise = flag is None or flag.raise_on_read_timeout
+            read_timeout = timeout.read_timeout if should_raise else 0.01
+            try:
+                data = await self.read_or_timeout(n, read_timeout)
+                break
+            except ReadTimeout:
+                if should_raise:
+                    raise
+
+        return data
 
     def write_no_block(self, data: bytes) -> None:
-        raise NotImplementedError()  # pragma: no cover
+        self.write_buffer += data
```

Comment:
I find this a bit misleading, because the `write_no_block` operation was influenced by asyncio's `.write()` / `.drain()` API, but what we're doing here doesn't *actually* write any data, and it won't until we make the next `.write()` call.

Should we repurpose this into `.queue_for_next_write()` or similar?